display_list/geometry/dl_region.cc

```diff
@@ -0,0 +1,260 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "flutter/display_list/geometry/dl_region.h"
+
+#include <cassert>
+
+namespace flutter {
+
+struct DlRegion::Span {
+  int32_t left;
+  int32_t right;
+};
+
+struct DlRegion::SpanLine {
+  int32_t top;
+  int32_t bottom;
+  // For performance reasons SpanLine must be trivially constructible.
+  // DlRegion is responsible for allocating and deallocating the
+  // memory for spans.
+  SpanVec* spans;
+
+  /// Inserts span into this span line. If there are existing spans
+  /// that overlap with the new span, they will be merged into single
+  /// span.
+  void insertSpan(int32_t left, int32_t right) {
+    SpanVec& spans = *this->spans;
+
+    auto size = spans.size();
+    for (size_t i = 0; i < size; ++i) {
+      Span& span = spans[i];
+      if (right < span.left) {
+        _insertSpanAt(i, left, right);
+        return;
+      }
+      if (left > span.right) {
+        continue;
+      }
+      size_t last_index = i;
+      while (last_index + 1 < size && right >= spans[last_index + 1].left) {
+        ++last_index;
+      }
+      span.left = std::min(span.left, left);
+      span.right = std::max(spans[last_index].right, right);
+      if (last_index > i) {
+        spans.erase(spans.begin() + i + 1, spans.begin() + last_index + 1);
+      }
+      return;
+    }
+
+    spans.push_back({left, right});
+  }
+
+  bool spansEqual(const SpanLine& l2) const {
+    SpanVec& spans = *this->spans;
+    SpanVec& otherSpans = *l2.spans;
+    assert(this != &l2);
+
+    if (spans.size() != otherSpans.size()) {
+      return false;
+    }
+    return memcmp(spans.data(), otherSpans.data(),
+                  spans.size() * sizeof(Span)) == 0;
+  }
+
+ private:
+  void _insertSpanAt(size_t index, int32_t x1, int32_t x2) {
+    spans->insert(spans->begin() + index, {x1, x2});
+  }
+};
+
+DlRegion::DlRegion() {
+  // If we can't memmove SpanLines addRect would be signifantly slower.
```

Comment:
... due to added costs of inserting and removing elements from the vector of lines?